File: Solutions/923. 3Sum With Multiplicity/923. 3Sum With Multiplicity.cpp
class Solution {
public:
    int threeSumMulti(vector<int>& arr, int target) {
        const int mod=1e9+7;
        unordered_map<int,long> um;
        for(auto& a:arr){
            ++um[a];
        }
        long res=0;
        for(auto& [k1,v1]:um){
            for(auto& [k2,v2]:um){
                int i=k1,j=k2,k=target-i-j;
                if(!um.count(k)){
                    continue;
                }
                if(i==j and j==k){
                    res+=um[i]*(um[i]-1)*(um[i]-2)/6;
                }else if(i==j and j!=k){
                    res+=um[i]*(um[i]-1)*um[k]/2;
                }else if(i<j and j<k){
                    res+=um[i]*um[j]*um[k];
                }
            }
        }
        return res%mod;
    }
};

File: Solutions/923. 3Sum With Multiplicity/923. 3Sum With Multiplicity.go
func threeSumMulti(arr []int, target int) int {
    m:=make(map[int]int)
    for _,a:=range(arr){
        m[a]+=1
    }
    res:=0
    for k1,_:=range(m){
        for k2,_:=range(m){
            i,j:=k1,k2
            k:=target-i-j
            if _,ok:=m[k];!ok{
                continue
            }
            if i==j && j==k{
                res+=m[i]*(m[i]-1)*(m[j]-2)/6
            }else if i==j && j!=k{
                res+=m[i]*(m[i]-1)*(m[k])/2
            }else if i<j && j<k{
                res+=m[i]*m[j]*m[k]
            }
        }
    }
    mod:=1e9+7
    return res%int(mod)
}